### PR TITLE
Task/hotword volume lowering

### PR DESCRIPTION
**Pull request by anthonyray**

### CODE REVIEW

File: snipssonos/use_cases/hotword/restore_volume.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+    def __init__(self, device_discovery_service, device_transport_control_service, state_persistence_service):
+        self.device_discovery_service = device_discovery_service
+        self.device_transport_control_service = device_transport_control_service
+        self.state_persistence_service = state_persistence_service
+
+    def process_request(self, request_object):
+        devices = self.state_persistence_service.get_all(Device)
+        if len(devices):
+            FIRST_DEVICE_VOLUME = devices[0].volume
```
Comment (paufabregat) [inline snipssonos/use_cases/hotword/restore_volume.py]: As you already did the effort to actually save the state of all the devices, it takes nothing to reset the volume to all of them with a loop here. It is not a big effort and already tackles a future feature.

File: action-sonos.py [hunk 2]
```diff
@@ -51,19 +52,43 @@
...
+
+        # We restore the volume to what it was before the hotword was detected.
+        for device in hermes.states.get('devices').values():
+            hermes.device_transport_control_service.set_volume(device)
+
+    return restore_volume_wrapper
+
+
 # Music management functions
+@restore_volume_for_hotword
```
Comment (paufabregat) [inline action-sonos.py:89]: Not completely sure of the magic behind this, I am assuming that the `restore_volume_for_hotword` function would be called always before the callback it sits on (in this case `addSong_callback`) and trigger the `restore_volume_wrapper` callback, right?

File: snipssonos/use_cases/volume/down.py [hunk 3]
```diff
@@ -2,18 +2,19 @@
...
-
     def process_request(self, request_object):
-        device = self.device_discovery_service.get()
-
-        device.decrease_volume(self.DEFAULT_VOLUME_DECREMENT)
+        devices = self.device_discovery_service.get_devices()
 
+        for device in devices:
+            device.decrease_volume(self.DEFAULT_VOLUME_DECREMENT)
         self.device_transport_control_service.volume_down(device)
```
Comment (paufabregat) [inline snipssonos/use_cases/volume/down.py]: this should be inside the for loop. Here you are just taking the last device that you iterate on in the loop.

File: snipssonos/use_cases/hotword/restore_volume.py (hunk 1 above)
Comment (paufabregat) [reply snipssonos/use_cases/hotword/restore_volume.py]: Also `FIRST_DEVICE_VOLUME ` should be lowercased, it is not a global.

File: snipssonos/use_cases/volume/down.py [hunk 4]
```diff
@@ -1,20 +1,28 @@
...
+        # We first check if the volume was lowered during hotword detection
+        if self.has_devices_persisted():
+            devices = self.state_persistence_service.get_all(Device)
+        else:
+            devices = self.device_discovery_service.get_devices()
 
         for device in devices:
             device.decrease_volume(self.DEFAULT_VOLUME_DECREMENT)
-        self.device_transport_control_service.volume_down(device)
+            self.device_transport_control_service.volume_down(device)
```
Comment (paufabregat) [inline snipssonos/use_cases/volume/down.py]: ah now I see, you fixed this :)

**Review by paufabregat — CHANGES_REQUESTED** [review]
Look super! Just requested for minimal changes and some clarification to understand some parts of the code.

**Comment by anthonyray** [comment]
I took your suggestions into account ;) ! 
Now merging the PR.

**Pull request merged**